scripts/receive-payjoin.py

```diff
@@ -4,19 +4,80 @@
 
 import sys
 from twisted.python.log import startLogging
-from jmbase import get_log, set_logging_level
-from jmclient import P2EPMaker, jm_single, load_program_config, \
-    WalletService, JMClientProtocolFactory, start_reactor, \
-    open_test_wallet_maybe, get_wallet_path, check_regtest, \
-    add_base_options
+from twisted.web.server import Site
+from twisted.internet import reactor
+import txtorcon
+from jmbase import get_log, set_logging_level, jmprint
+from jmclient import jm_single, load_program_config, \
+    WalletService, open_test_wallet_maybe, get_wallet_path, check_regtest, \
+    add_base_options, PayjoinServer
 from jmbase.support import EXIT_FAILURE, EXIT_ARGERROR
-from jmbitcoin import amount_to_sat
+from jmbitcoin import amount_to_sat, amount_to_btc, encode_bip21_uri, \
+     CCoinAddress
+
+# txtorcon outputs erroneous warnings about hiddenservice directory strings,
+# annoyingly, so we suppress it here:
+import warnings
+warnings.filterwarnings("ignore")
 
 jlog = get_log()
 
-def receive_payjoin_main(makerclass):
+bip78_amount = None
+bip78_receiving_address = None
+
+def setup_failed(arg):
+    jlog.error("SETUP FAILED", arg)
+    reactor.stop()
+
+def create_onion_ep(t, hs_public_port):
+    return t.create_onion_endpoint(hs_public_port)
+
+def onion_listen(onion_ep, site):
+    return onion_ep.listen(site)
+
+def print_host(ep):
+    """ Callback fired once the HS is available;
+    receiver user needs a BIP21 URI to pass to
+    the sender:
+    """
+    jmprint("Your hidden service is available. Please now pass "
+            "this URI string to the sender to effect the payjoin "
+            "payment:")
+    jmprint(bip21_uri_from_onion_hostname(str(ep.getHost().onion_uri),
+                                          ep.getHost().onion_port))
+    jmprint("Keep this process running until the payment is received.")
+
+def bip21_uri_from_onion_hostname(host, port):
+    """ Encoding the BIP21 URI according to BIP78 specifications,
+    and specifically only supporting a hidden service endpoint.
+    Note: we hardcode http; no support for TLS over HS.
+    Second, note we convert the global amount-in-sats bip78_amount
+    to BTC denomination as expected by BIP21.
+    """
+    full_pj_string = "http://" + host + ":" + str(port)
```

Comment:
Port should not be added to pj callback url if it's 80.

Reply:
Perhaps, in line with the above conversation, I was worried that we choose a port that's conflicting. I don't really want a bunch of extra code to handle that case, but maybe it's fine, if the hidden service fails to start it should be picked up as "setup failed" in the errback.

Reply:
It should not conflict, as the new unique hidden service is generated each time.

Reply:
But it can't serve on the same port if there's already another web server running on it.

Reply:
Why would another web server already be running on a freshly generated onion service?

Reply:
Another web server on the same port on the same operating system, I mean; that would conflict.

Reply:
Oh, are you saying that doesn't conflict, because it's Tor?

Reply:
Yes, it does not. Webserver, like nginx, does not know anything about Tor by itself. And different onion services does not affect each other.

Reply:
I see. So more reason to just say 80 default, I will make the edits soon.

Reply:
Addressed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/682/commits/06d8772b366fba84f63e252c772c465219376b9d